### ISSUE TITLE
ImageRecordReader should accept ordered labels

### ISSUE THREAD

**Issue by crockpotveggies** [deeplearning4j/DataVec#222]
Currently, if we want to split a dataset for train/test, the following method is suggested:
```
BalancedPathFilter balancedFilter = new BalancedPathFilter(rng, NativeImageLoader.ALLOWED_FORMATS, labelMaker, 330000, 8100, 1,30);
split = fileSplit.sample(balancedFilter, 0.8, 0.2);
RecordReader trainRR = new ImageRecordReader(height, width, channels, labelMaker, flip);
    trainRR.initialize(rrConf, split[0]);
RecordReader testRR = new ImageRecordReader(height, width, channels, labelMaker, flip);
    testRR.initialize(rrConf, split[1]);
```

The problem is that in practice, this leads to a significant reduction in training data since BalancePathFilter removes a large number of paths from the final training set. Using the alternative, RandomPathFilter, creates a whole new issue further in composing a training process since the file split will not contain the same number of labels:

```
RecordReader testRR = new ImageRecordReader(height, width, channels, labelMaker, flip);
    testRR.initialize(rrConf, split[1]);
new RecordReaderDataSetIterator(testRR, batchSize, 1, trainRR.getLabels().size());
```

When manually specifying the number of classes, this does not guarantee that label ordering will be the same. This incorrect ordering will produce incorrect results. However, if the user wants to use BalancedPathFilter, their dataset is drastically reduced in size, meaning that the maximum potential of the net is not achieved.

Proposed method of fixing this is manually specifying labels via a new signature for initialize method such as `testRR.initialize(rrConf, split[1], trainRR.getLabels())`.

Internally, that should fix a label ordering problem.

**Comment by crockpotveggies** [comment]
Discovered that BaseImageRecordReader has a setting in its base class for setting labels:

```
    public void initialize(Configuration conf, InputSplit split) throws IOException, InterruptedException {
        this.appendLabel = conf.getBoolean(APPEND_LABEL, false);
        this.labels = new ArrayList<>(conf.getStringCollection(LABELS));
```

However, this is unfriendly for UX since this requires the user to create a string collection and pass it to a hidden config. I believe adding this to a new constructor will be much more effective and user friendly.

**Comment by saudet** [comment]
I thought Alex already added something somewhere to sort the labels ... @AlexDBlack ?

**Comment by AlexDBlack** [comment]
Yup, labels are sorted alphabetically:
https://github.com/deeplearning4j/DataVec/blob/master/datavec-data/datavec-data-image/src/main/java/org/datavec/image/recordreader/BaseImageRecordReader.java#L150

However, this can fail if the test set doesn't have examples for some of the classes (as the set of possible labels, and hence label/index assignment are done by scanning the set of possible files).

There's a setLabels method too:
https://github.com/deeplearning4j/DataVec/blob/master/datavec-data/datavec-data-image/src/main/java/org/datavec/image/recordreader/BaseImageRecordReader.java#L311

I wouldn't be opposed to a new initialize method - only downside there is that if we do ```RecordReader rr = new ImageRecordReader(...)``` then the initialize method isn't visible (without casting). Same with the setLabels method though, so...

**Comment by crockpotveggies** [comment]
In the meantime, the workaround when a constructor is not present is to use this code:

```
    ImageRecordReader trainRR = new ImageRecordReader(height, width, channels, labelMaker, combinedTransform);
    trainRR.initialize(trainData);
    ImageRecordReader testRR = new ImageRecordReader(height, width, channels, labelMaker, combinedTransform);
    testRR.setLabels(trainRR.getLabels());
    testRR.initialize(testData);
```

**Comment by saudet** [comment]
What about if BalancedPathFilter created cycles over all examples of each classes, repeating the cycles of each class independently, until the class with the largest number of examples completes at least one cycle?

**Comment by crockpotveggies** [comment]
@saudet not a bad idea, what kind of performance/timing would that take?
